### ISSUE TITLE
new_record? returns false after rollback if record is created and then updated in the transaction

### ISSUE THREAD

**Issue by wildfiler** [rails/rails#22066]
The following code fails in rails > 4.0 and in master. It works fine for 3.2.22.

For the second call `new_record?` returns correct value, but `id` is still present for this record.

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', github: 'rails/rails'
  gem 'arel', github: 'rails/arel'
  gem 'rack', github: 'rack/rack'
  gem 'sprockets', github: 'rails/sprockets'
  gem 'sprockets-rails', github: 'rails/sprockets-rails'
  gem 'sass-rails', github: 'rails/sass-rails'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_rollback_restore_state_on_object_created_and_updated_in_transaction
    Post.transaction do
      @post = Post.create!
      @post.touch
      raise ActiveRecord::Rollback
    end

    assert @post.new_record?
  end

  def test_new_record_state_dont_changes_for_sequential_calls
    Post.transaction do
      @post = Post.create!
      @post.touch
      raise ActiveRecord::Rollback
    end

    assert_equal @post.new_record?, @post.new_record?
  end
end
```

Update: can be related to #13744

**Comment by repinel** [comment]
This can be fixed by #21020

**Comment by louis-she** [comment]
I'm still facing this problem in Rails 5 : (
